### ISSUE TITLE
Attach PipelineId to all ScriptMsg

### ISSUE THREAD

**Issue by paulrouget** [servo/servo#17201]
Many ScriptMsg already require a PipelineId field.

Many of these messages are sent to the constellation to be then sent to the embedder via the compositor (status, favicon, loadcomplete, loadstate, etc). And as of now, the decision to send the message to the embedder (which usually is either the pipeline the top level one or not) is done in on of these 3 places: in /script/, in compositor.rs, or in constellation.rs.

I think the logic of forwarding the event or not should all happen in constellation.rs. The constellation has more context than /script/, and we eventually want to get rid of the compositor step (#15934).

For the constellation to make the correct decision of forwarding the event, it needs the PipelineId.

I'm looking for a way to easily attach the PipelineId to all the ScriptMsg.

I was thinking about change the ScriptMsg type to:

```rust
struct ScriptMsg {
  source: PipelineId,
  message: ScriptMsgContent,
}

enum ScriptMsgContent {
  SetPageTitle,
  Focus,
  …
}
```

… but it's kind of cumbersome. And we don't always have an easy access to the pipeline_id. Would there be a better way to do that?

**Comment by paulrouget** [comment]
I want also thinking about creating a struct `CompositorChannel`:

```rust
struct CompositorChannel {
  sender: IpcSender<ScriptMsg>,
  pipeline_id: PipelineId,
}

impl CompositorChannel {
  fn send(msg) {
    self.sender.send(self.pipeline_id),
  }
  fn clone_with_new_pipeline(pipeline_id) -> CompositorChannel {
    CompositorChannel {
      pipeline_id: pipeline_id,
      send: self.sender.clone(),
    }
  } 
}
```

… and instead of calling `constellation_chan.clone()` we would use `constellation_chan.clone_with_new_pipeline(pipeline_id)`.

**Comment by asajeffrey** [comment]
Do you mean:
```rust
   struct CompositorChannel {
      sender: Sender<(PipelineId, ScriptMsg)>,
      pipeline_id: PipelineId,
   }
```
If so, apart from the naming, this seems sensible. I used a similar trick in worklets: https://github.com/servo/servo/blob/0934df77fdbddcb6ae1534e3c1a150272aea73b2/components/script/dom/worklet.rs#L657

**Comment by paulrouget** [comment]
oops, yes, it's what I meant.

Thanks.

**Comment by paulrouget** [comment]
Fixed by #17425
